nri-redis/src/Redis/Mock.hs

```diff
@@ -106,299 +105,413 @@ expectInt val =
+    Model
+      { hash = HM.empty,
+        scans = Array.initialize 1 (always HM.empty)
+      }
 
-doQuery ::
-  Internal.Query a ->
-  HM.HashMap Text RedisType ->
-  (HM.HashMap Text RedisType, Result Internal.Error a)
-doQuery query hm =
-  case query of
-    Internal.Apply fQuery xQuery ->
-      let (hm1, f) = doQuery fQuery hm
-          (hm2, x) = doQuery xQuery hm1
-       in (hm2, map2 (\f' x' -> f' x') f x)
-    Internal.Del keys ->
-      List.foldl
-        ( \key (hm', count) ->
-            if HM.member key hm'
-              then (HM.delete key hm', count + 1)
-              else (hm', count)
-        )
-        (hm, 0 :: Int)
-        (NonEmpty.toList keys)
-        |> Tuple.mapSecond Ok
-    Internal.Exists key ->
-      ( hm,
-        Ok (HM.member key hm)
-      )
-    Internal.Expire _ _ ->
-      -- Expiring is an intentional no-op in `Redis.Mock`. Implementing it would
-      -- likely be a lot of effort, and only support writing slow tests.
-      ( hm,
-        Ok ()
-      )
-    Internal.Get key ->
-      ( hm,
-        HM.lookup key hm
-          |> Prelude.traverse expectByteString
-      )
-    Internal.Getset key value ->
-      ( HM.insert key (RedisByteString value) hm,
-        HM.lookup key hm
-          |> Prelude.traverse expectByteString
-      )
-    Internal.Hdel key fields ->
-      case HM.lookup key hm of
-        Nothing ->
-          ( hm,
-            Ok 0
-          )
-        Just (RedisHash hm') ->
-          let hmAfterDeletions = Prelude.foldr HM.delete hm' fields
-           in ( HM.insert key (RedisHash hmAfterDeletions) hm,
-                HM.size hm' - HM.size hmAfterDeletions
-                  |> Prelude.fromIntegral
-                  |> Ok
-              )
-        Just _ ->
-          ( hm,
-            Err wrongTypeErr
-          )
-    Internal.Hget key field ->
-      case HM.lookup key hm of
-        Nothing ->
-          ( hm,
-            Ok Nothing
-          )
-        Just (RedisHash hm') ->
-          ( hm,
-            Ok (HM.lookup field hm')
-          )
-        Just _ ->
-          ( hm,
-            Err wrongTypeErr
-          )
-    Internal.Hgetall key ->
-      ( hm,
-        HM.lookup key hm
-          |> Prelude.traverse expectHash
-          |> map
-            ( \res ->
-                case res of
-                  Just hm' -> HM.toList hm'
-                  Nothing -> []
+doQuery :: Internal.Query a -> Model -> (Model, Result Internal.Error a)
+doQuery query model =
+  let hm = hash model
+      updateHash = \h -> model {hash = h}
+   in case query of
+        Internal.Apply fQuery xQuery ->
+          let (model1, f) = doQuery fQuery model
+              (model2, x) = doQuery xQuery model1
+           in (model2, map2 (\f' x' -> f' x') f x)
+        Internal.Del keys ->
+          List.foldl
+            ( \key (hm', count) ->
+                if HM.member key hm'
+                  then (HM.delete key hm', count + 1)
+                  else (hm', count)
```

Comment:
GitHub is rendering this diff in the most horrible way. Not sure what I can do about it.
Basically all branches now need to update the whole `model` instead of just `hm`.